[estelle] iP

estellelim opened the pull request:
### Muffin
> "Your mind is for having ideas, not holding them." - David Allen ([source](https://dansilvestre.com/productivity-quotes))

Muffin frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- responds fast

All you need to do is,

1. download it.
2. add you tasks.
3. let it manage your tasks for you 😉

And it is **FREE**!

Features: 

- [x] Managing tasks
- [x] Managing deadlines
- [x] Managing events
- [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[] args) {
        new Muffin().run();
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

estellelim committed:
Add skeletal version

estellelim committed:
Add skeletal version

estellelim committed:
Add echo

estellelim committed:
Add actions

estellelim committed:
Add mark & unmark feature

estellelim committed:
Add types of tasks

estellelim committed:
Add Automated Text Ui Testing

estellelim committed:
Add exceptions

estellelim committed:
Add delete feature

estellelim committed:
Add save tasks feature

estellelim committed:
Add date format for Deadline

estellelim committed:
Add date format for Event task

estellelim committed:
Merge branch 'branch-Level-8'
Merging the date format feature onto master.

estellelim committed:
Extract more classes

estellelim committed:
Add Package

estellelim committed:
Add package

estellelim committed:
Merge branch 'add-gradle-support'
Add gradle support.

estellelim committed:
Add gradle

estellelim committed:
Add JUnit tests

estellelim committed:
Add JavaDoc comments

estellelim committed:
Add find feature

estellelim committed:
Merge pull request #1 from estellelim/branch-A-JavaDoc

Add JavaDoc comments

estellelim committed:
Merge pull request #2 from estellelim/branch-Level-9

Add find feature

estellelim committed:
Merge branch 'branch-A-CodingStandard'
Merge code.

estellelim committed:
Merge branch 'branch-Level-9'
Merge find feature.

beginner-web-developer commented on src/main/java/muffin/Event.java:
```diff
@@ -0,0 +1,41 @@
+
+    @Override
+    public String toString() {
+        String f = from.format(DateTimeFormatter.ofPattern("E, MMM d yyyy"));
+        String t = to.format(DateTimeFormatter.ofPattern("E, MMM d yyyy"));
```
Maybe you can consider using more descriptive names, such as fromDate and toDate.

beginner-web-developer commented on src/main/java/muffin/FileProcessor.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    public void writeToFile(String filePath, ArrayList<Task> tasks) {
+        try {
+            ArrayList<String> content = new ArrayList<>();
```
The arraylist contains many String objects, so maybe a more approriate naming for this variable would be `contents`

beginner-web-developer commented on src/main/java/muffin/Muffin.java:
```diff
@@ -0,0 +1,178 @@
+    /**
+     * An instance of FileProcessor to handle reading and writing tasks to a file.
+     */
+    FileProcessor fp = new FileProcessor();
```
It may be more suitable for this variable to have a more descriptive name, as it is used throughout the whole program. Maybe a better name can be `fileProcessor`

beginner-web-developer commented on src/main/java/muffin/TaskList.java:
```diff
@@ -0,0 +1,114 @@
+    /**
+     * The list of tasks.
+     */
+    ArrayList<Task> list;
```
Maybe `tasks` would be a better name for this variable, to be more descriptive. Same for the `ArrayList<Task>` at line 95.

beginner-web-developer reviewed (commented):
Overall you did a nice job in the naming of the variables, but there are just a few nits to consider.

estellelim committed:
Add checkstyle and fixed a few style violations

estellelim committed:
Merge branch 'branch-A-CheckStyle'
Merge Checkstyle changes to master.

Rachael-Chan commented on src/main/java/muffin/Muffin.java:
```diff
@@ -0,0 +1,178 @@
+                throw new MuffinException("Um... Not sure what you mean...");
+            }
+
+            if (command == Command.MARK || command == Command.UNMARK || command == Command.DELETE) {
+                try {
+                    int index = Integer.parseInt(parts[1]);
+                    if (index > len) {
+                        throw new MuffinException("Oh no! There is only " + len + " tasks.");
+                    } else if (index < 1) {
+                        throw new MuffinException("Oh no! This task does not exist!");
+                    }
+                } catch (NumberFormatException e) {
+                    throw new MuffinException("Oh no! You must state the number of the task you'd like to edit!");
+                }
+            }
```
Nice use of enums

Rachael-Chan reviewed (commented):
LGTM - overall, great job in adhering to coding standards

estellelim committed:
Add GUI

estellelim committed:
The code currently does not document key assumptions explicitly.

It needs to change because it’s essential to ensure assumptions are clear and verified during development and runtime.

Using `assert` ensures that potential issues are caught early and helps developers maintain a clear understanding of what is expected at different stages of the program.

Ensure that assertions are enabled during the runtime for the checks to be effective.

estellelim committed:
Refactor readFromFile and writeToFile methods in FileProcessor for improved clarity and maintainability

- Refactored task creation logic in readFromFile by moving it to the new TaskFactory class.
  - Centralized task creation using factory methods to eliminate switch-case logic in readFromFile.
  - Simplified handling of task types (Todo, Deadline, Event) and status marking.

- Enhanced writeToFile method by using polymorphism.
  - Each Task subclass now defines its own toFormattedString method for cleaner code and type-specific formatting.
  - Removed instanceof checks, adhering to better OOP practices.

- Added detailed Javadoc comments for both methods and supporting classes.
- Improved overall code maintainability and readability, making it easier to extend task types in the future.

estellelim committed:
Refactor array concatenation to use Streams

Replaced manual arraycopy logic with Stream API for more readable and concise code.

estellelim committed:
Merge pull request #3 from estellelim/branch-A-Assertions

Branch a assertions

estellelim committed:
Merge pull request #4 from estellelim/branch-A-CodeQuality

Refactor readFromFile and writeToFile methods in FileProcessor for im…

estellelim committed:
Merge pull request #5 from estellelim/branch-A-Streams

Refactor array concatenation to use Streams

estellelim committed:
Implement undo functionality for task operations.

Currently, the program does not have support for undoing task actions. This restricts user flexibility when managing tasks, as they cannot revert their last operation.

Introducing an undo feature with a stack that tracks task operations in the TaskList class will allow users to revert their last action. The stack stores each action with relevant task data and position.

This method was chosen to centralize task operation history in TaskList, maintaining modularity and ensuring that task operations and their undo logic are managed in the same place. File updates are triggered after each action and undo, ensuring file consistency.

estellelim committed:
Add JavaDoc comments

estellelim committed:
Merge pull request #6 from estellelim/C-Undo

C undo

estellelim committed:
Add JavaDoc comment

estellelim committed:
Add missing JavaDoc comments

estellelim committed:
Change UI color

estellelim committed:
Add user guide in README

estellelim committed:
Edit path to screenshot

estellelim committed:
Add automatic app exit by command bye

estellelim committed:
Edit code to coding standard

estellelim committed:
Update the README

estellelim committed:
Add detailed error messages
